TSFC integration

FabioLuporini opened the pull request:
This PR does two things:

- fixes the integration with TSFC and enables expression rewriting
- drastically improves the quality of the codebase

TODO:

- integration of padding + data alignment

Comments welcome 

There are a few pending issues and the need for a little bit of rebasing, but for the rest it should be pretty much ready

dorugeber commented:
Ugh.  Obvious comment: the commit is +2400/-2300, and there's lots of 'new files added' and 'old files deleted' which I assume is just code shuffling around.  If you actually want someone to read the PR in detail (maybe you don't!), is there a sane way to split this into a trivial "shuffling" PR and a PR where the changes happen?

dorugeber commented:
(If it would take you an hour, just do it.  If it would take you a week, don't!)

FabioLuporini commented:
That's true. It's actually just code refactoring. The only real new stuff is in cse.py, plus a few things in hoister.py. The rest is just code move/renamed/refactored.

Thanks

miklos1 commented on coffee/cse.py:
```diff
@@ -0,0 +1,407 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
Wrong year, and possibly wrong copyright holder.

FabioLuporini replied:
fixed the year

miklos1 commented on coffee/cse.py:
```diff
@@ -0,0 +1,407 @@
+
+
+class Temporary():
+
```
Docstring?

FabioLuporini replied:
added

miklos1 commented on coffee/cse.py:
```diff
@@ -0,0 +1,407 @@
+
+
+class CSEUnpicker():
+
```
Docstring?

FabioLuporini replied:
added

miklos1 commented on coffee/cse.py:
```diff
@@ -0,0 +1,407 @@
+        return best
+
+    def unpick(self):
+        fors = visit(self.header, info_items=['fors'])['fors']
```
Docstring?

FabioLuporini replied:
just the main cseunpicker function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

miklos1 commented on coffee/cse.py:
```diff
@@ -0,0 +1,407 @@
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
A module docstring would be nice as well.

miklos1 commented on coffee/expander.py:
```diff
@@ -0,0 +1,244 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
copy-paste

FabioLuporini replied:
fixed

wence- commented on coffee/expander.py:
```diff
@@ -0,0 +1,244 @@
+                self._map.pop(i)
+                if self._hits[i] > 0:
+                    was_hit = True
+        return was_hit
```
AKA:
```python
def invalidate(self, exp):
    for i, j in self._map.items():
        if str(j) = str(exp):
            self._map.pop(i)
            if self._hits[i] > 0:
                return True
     return False
```

FabioLuporini replied:
true :) changed

wence- commented on coffee/rewriter.py:
If I read the docs for pulp correctly, you're meant to pull the answer and the objective out of the solved problem.  Does it also update variables "in place"?

wence- commented on coffee/rewriter.py:
Also, if the order is crucial is it safe to iterate over a dict and collect things?

miklos1 commented on coffee/expander.py:
```diff
@@ -0,0 +1,244 @@
+        else:
+            raise RuntimeError("Expansion error: unknown node: %s" % str(node))
+
+    def expand(self, should_expand, **kwargs):
```
docstring?

FabioLuporini replied:
just the main expander function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

miklos1 commented on coffee/expression.py:
```diff
@@ -155,16 +172,11 @@ def mode(self, value):
-        * mode: the suggested rewrite mode.
-    """
+    ``kwargs``. ``kwargs`` accepts the following keys: parent, loops_info,
+    linear_dims, mode."""
```
Why did you remove the explanation of `parent`, `loops_info` etc. ?

FabioLuporini replied:
it was basically the same description as the one you can find in the MetaExpr class' docstring

miklos1 commented on coffee/factorizer.py:
```diff
@@ -0,0 +1,233 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
copy-pasta

FabioLuporini replied:
fixed

miklos1 commented on coffee/factorizer.py:
```diff
@@ -0,0 +1,233 @@
+
+
+class Factorizer():
+
```
docstring?

FabioLuporini replied:
just the main factorizer function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

miklos1 replied:
meh... how am I supposed to know that? Maybe add some reference at least.

FabioLuporini replied:
ok doing it now

miklos1 commented on coffee/factorizer.py:
```diff
@@ -0,0 +1,233 @@
+        else:
+            return Term([], [node])
+
+    def factorize(self, should_factorize, **kwargs):
```
docstring?

FabioLuporini replied:
just the main factorizer function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

FabioLuporini replied:
ok done now

miklos1 commented on coffee/hoister.py:
```diff
@@ -0,0 +1,431 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
copy-pasta

miklos1 commented:
This patch removes 22 `dcopy`s and adds 13. I guess that is an improvement.

FabioLuporini commented on coffee/rewriter.py:
Yes. I found it in some tutorials tbh

As for the order: what I actually meant is that all ``nodes`` have to come before all ``other_nodes``

FabioLuporini commented:
Addressed most of your comments

Please pull

Now the zero-elimination engine is integrated. As a temporary experiment, I've added this transformation to the default optimisation pipeline (i.e., O2) . @wence-  @miklos1  Can you please beta-test this branch at least for functionality ? I'm still running a few experiments for performance. 

Thanks

miklos1 commented:
Did you just fix -O4? The hyperelasticity test is unexpectedly passing again.

wence- commented:
So that xfail was on old pre-tsfc coffee.  I can entirely imagine that many things have changed.  For example, if no optimisations occur then the test xpasses.

miklos1 commented:
With last week's `tsfc-updates` COFFEE, it was still xfailing, but now it passes, and COFFEE claims some flop reduction (though not as much as with -O2).

FabioLuporini commented:
I've been changing the transformations applied by each optimisation level. 

Now -O4 does what previously -O3 did. Some fixes went it and made it work. 

Also (see last commit in capital letters), at the moment O2 == O3. I've temporarily disabled alignment and padding in {O2,O3} because most users use gcc, and gcc doesn't like auto-vectorising the code, despite all my efforts trying to generate "easily vectorisable source code". Alignment and padding are applied instead at level O4.

I was thinking that one way of deciding whether to apply padding and data alignment is to check whether the underlying compiler is "recent enough" to support the #pragma simd directive (which enforces vectorisation). Most recent versions of gcc have it. We could think of spoofing somehow the compiler version...

FabioLuporini commented:
Added logic for avoiding zero elimination 

Addressed your last comments concerning docstrings

Tests are now running. Maybe it is also now a good time to purge COFFEE from crap ``Par`` nodes :) a git grep seems to show they are never used to construct objects...

kynan commented on .travis.yml:
```diff
@@ -10,6 +10,11 @@ language: python
+addons:
+  apt:
+    packages:
+    - glpk-utils
```
Is this going to prevent Travis from running on container infrastructure?

FabioLuporini replied:
I honestly have no idea... @miklos1  ?

miklos1 replied:
I mostly added this as documentation, because COFFEE now relies on this, even though its test suite does not actually uses it. I have no idea about Travis.

kynan commented on coffee/base.py:
```diff
@@ -596,8 +601,13 @@ def rvalue(self, val):
         self.children[1] = val
 
     def gencode(self, not_scope=False):
```
Double negation? `not_scope=False` == `scope=True`?

FabioLuporini replied:
yes that's bad... I should maybe change it

kynan commented on coffee/cse.py:
```diff
@@ -0,0 +1,419 @@
+from expression import MetaExpr
+
+
+class Temporary():
```
New style class?

FabioLuporini replied:
do you have a preference ?

miklos1 replied:
New style class is always better.

FabioLuporini replied:
OK

miklos1 commented:
Purged `Par` nodes, so this pull request now depends on firedrakeproject/firedrake#746.

kynan commented on coffee/cse.py:
This does't do what you expect it to

wence- commented on coffee/cse.py:
By which he means you wanted `for i in bounds: assert i in levels.keys()`

wence- commented on coffee/cse.py:
Better to write: `assert set(levels.keys()) == set(bounds)` I think.

wence- commented on coffee/cse.py:
It's unclear why you would want an ordered dict here given that levels is an unordered dict.

FabioLuporini commented on coffee/cse.py:
uooops! Fixing

FabioLuporini commented on coffee/cse.py:
good spot

FabioLuporini commented:
Addressed latest comments (tests running now)

kynan commented on coffee/expander.py:
```diff
@@ -0,0 +1,244 @@
+            retval = FindInstances.default_retval()
+            exp = [s for s in FindInstances(Symbol).visit(exp, ret=retval)[Symbol]
+                   if s.symbol in self.hoisted and self.should_expand(s)][0]
+        except:
```
How would the above fail?

FabioLuporini replied:
[0] at the end

miklos1 replied:
This is very loose, because `except:` will catch expression thrown from anywhere inside this expression. Can you make it more specific?

wence- replied:
So you're checking if len(...) == 0?

FabioLuporini replied:
yes. OK, I'll make it more specific

FabioLuporini replied:
I did better than that, I deleted a whole chunk of code that was only useful in a negligible number of cases

kynan commented on coffee/hoister.py:
```diff
@@ -0,0 +1,431 @@
+from utils import *
+
+
+class Extractor():
```
New style class?

kynan commented on coffee/factorizer.py:
```diff
@@ -0,0 +1,237 @@
+        return Term(operands, factors, op)
+
+
+class Factorizer():
```
New style class?

kynan commented on coffee/factorizer.py:
```diff
@@ -0,0 +1,237 @@
+from utils import *
+
+
+class Term():
```
New style class?

kynan commented on coffee/hoister.py:
```diff
@@ -0,0 +1,431 @@
+        return (dep_n, self.STOP)
+
+
+class Hoister():
```
New style class?

kynan commented on coffee/hoister.py:
```diff
@@ -0,0 +1,431 @@
+class Hoister():
+
+    # How many times the hoister was invoked
+    _handled = 0
```
Suppose this functionality could be factored into a common base class?

kynan commented on coffee/rewriter.py:
```diff
@@ -31,18 +31,20 @@
 
 
-class ExpressionRewriter(object):
+class ExpressionRewriter():
```
Why make this an old style class?

kynan commented on coffee/utils.py:
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
+###############################################################
+
+
+class StmtTracker(OrderedDict):
```
I suppose a `StmtTracker` HAS-A `OrderedDict` rather than IS-A `OrderedDict`?

FabioLuporini replied:
IS-A

wence- replied:
Confusingly, it both *has* an ordereddict and *is* an ordereddict :).

FabioLuporini replied:
Ah right I had forgotten about it. It's because I need access by value too at some point

kynan commented on coffee/utils.py:
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
+        * The parent of the closest loop
+    """
+
+    class StmtInfo(object):
```
Use a [`namedtuple`](https://docs.python.org/2/library/collections.html#collections.namedtuple) instead?

FabioLuporini replied:
Ah, I didn't know about it

kynan commented on requirements.txt:
```diff
@@ -1 +1,2 @@
 networkx
+PuLP
```
Does COFFEE work with the latest PuLP release?

FabioLuporini replied:
The latest is 1.6.1 and it does

wence- commented on coffee/utils.py:
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, self.StmtInfo):
+            if not isinstance(value, tuple):
```
`if not isinstance(value, (tuple, self.StmtInfo))`

wence- replied:
Oh, I see why not.

wence- commented on coffee/utils.py:
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
+        if not isinstance(value, self.StmtInfo):
+            if not isinstance(value, tuple):
+                raise RuntimeError("StmtTracker accepts tuple or StmtInfo objects")
+            value = self.StmtInfo(**dict(zip(self.StmtInfo.INFO, value)))
```
`zip` throws away entries from the longer list.  Do you want to assert that `len(value) == len(self.StmtInfo.INFO`?

FabioLuporini replied:
adding the check

FabioLuporini commented:
OK, again, addressed your comments.

miklos1 commented:
So it seems that COFFEE only works in addition to the "unroll `IndexSum`s" TSFC optimisation. If that optimisation is disabled (it is on by default), then COFFEE fails. This is not blocker, but it is nice to have it documented.

wence- commented:
fails, or does nothing?

miklos1 commented:
Fails.

miklos1 commented:
firedrakeproject/firedrake#746 is merged. @FabioLuporini: get this ready for the merge.
